### PR TITLE
[#5050] Create TrailingBody cops for Module & Class

### PR DESCRIPTION
**Pull request by garettarrowood**
This PR is based on the discussion on #5050. This PR:

1. Refactors corrector logic into a `LineBreakCorrector`
1. Creates new `Style/TrailingBodyOnClassDefinition`
1. Creates new `Style/TrailingBodyOnModuleDefinition`
1. Refactors new/old logic into a `TrailingBody` mixin.

While this works, it raises a lot of questions. I'm very receptive to refactoring/renaming whatever is suggested. Questions/concerns:

1. Classes and modules do indeed have `body`s.
```ruby
def body
  node_parts[index]
end
```
Why is ^ duplicated in so many `_node.rb`s?  `node_parts` is also duped _many_ times. Is there a rational behind this, or is it just how things unfolded?  In a separate PR, would it be ok to move  `node_parts` and `body` into `node.rb`?

2. `ClassDefinition` and `ModuleDefinition` do not sound right to me. But what is a better alternative... `*Statement`, `*Declaration`?  Since classes and modules may be reopened, `*Definition` does not seem to fit. WDYT?

Ready for review and suggestions. I expect this needs some serious tweaks, but wanted to get feedback.

-----------------

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/trailing_body_on_class_definition.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for trailing code after the class definition.
+      #
+      # @example
+      #   # bad
+      #   Class Foo; def foo; end
```
Comment (bbatsov) [inline lib/rubocop/cop/style/trailing_body_on_class_definition.rb]: Class -> class

File: lib/rubocop/cop/correctors/line_break_corrector.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # This class manipulates line breaks for autocorrection
```
Comment (bbatsov) [inline lib/rubocop/cop/correctors/line_break_corrector.rb]: That's a bit vague. :-) I also don't think we've used the term "line break" anywhere in the code so far.

File: lib/rubocop/cop/style/trailing_body_on_class_definition.rb (hunk 1 above)
Comment (garettarrowood) [reply lib/rubocop/cop/style/trailing_body_on_class_definition.rb]: fixed in the final commit

Comment (garettarrowood) [reply lib/rubocop/cop/style/trailing_body_on_class_definition.rb]: I will clean it up.  But want to get your thoughts on Cop names, and any other naming in here beforehand.

**Comment by bbatsov** [comment]
> Why is ^ duplicated in so many _node.rbs? node_parts is also duped many times. Is there a rational behind this, or is it just how things unfolded? In a separate PR, would it be ok to move node_parts and body into node.rb?

That's just some historical artefact. I don't recall any particular reasoning behind this, so we can improve it. 

> ClassDefinition and ModuleDefinition do not sound right to me. But what is a better alternative... *Statement, *Declaration? Since classes and modules may be reopened, *Definition does not seem to fit. WDYT?

I'd use "In" instead of "On" if we stick to the current names, which might be very correct, but I'm fine with them. I'm also fine with `ClassStatement` or even just `Class`.

**Comment by garettarrowood** [comment]
Thanks! I went with just `Style/TrailingBodyOnClass` and `Style/TrailingBodyOnModule`.

I'll finish off the autocorrection needed for #5050 in a new PR, then DRY up our `node_parts` and `body` methods in 1 or 2 more.

**Comment by Drenmi** [comment]
> `node_parts` is also duped many times.

If I recall correctly, there was another `#node_parts` method, so this was to avoid confusion. I also envisioned having this method act as living documentation of the different node parts inside the destructuring method. Example for a `case` node:

```ruby
def node_parts
  condition, *when_branches, else_branch = to_a
end
```

I never got around to checking if there was a noticeable performance penalty for it though, so it remained a plain alias. I guess I'm okay with this moving to `Node`.

> Why is ^ duplicated in so many `_node.rb`s?

Because not all node types with bodies will have the body in the same sequence. Furthermore I'm not sure if there's a strong enough conceptual connection between different nodes that have bodies to make a common abstraction helpful. The other abstractions, e.g. `ConditionalNode`, were made because it proved helpful in refactoring existing cops. (I'm generally quite cautious about choosing the wrong abstraction.)

Another trade-off is that moving away the named methods for the different parts from the extension makes it harder to get to what the node's destructuring looks like.

What I am sure about is it should not go in `Node`, since most node types don't have a body.

**Comment by garettarrowood** [comment]
Thanks for this helpful clarification @Drenmi !  A PR to move `node_parts` to `Node` has already been merged (#5482 ). But given this history, I will not make any further changes in regards to `body`.

If you have second thoughts about even moving `node_parts`, feel free to comment on that PR and I'll prep a revert.  Just having `node_parts` will let allow me to write some more understandable code in the new `TrailingBody` mixin.

**Pull request merged**